### ISSUE TITLE
Realtime messages not being delivered

### ISSUE THREAD

**Issue by agiacomolli** [ably/ably-go#619]
We're experiencing an issue with realtime messages not being delivered end-to-end. Publishing the message from one side (web app) to another (Go client) and sometimes the Go client does not react. Ably logs on the Go client side did not show any info about a message coming though from the web app. This happens consistently on the same few devices and eventually on the others.

The strange thing is that it seems to be more reliable if we listen to the same channel with another client. We first observed it with Ably dev console listening to the same channel Go client listens to. Only a few messages published from web app are not delivered to the Go client.

We recently migrate from ably-go `1.2.12` to `1.2.13`, but revisiting logs we can confirm this occurs on later versions.

Any tips appreciated.

**Comment by stmoreau** [comment]
Hi @agiacomolli, and thanks for raising your issue with us. Can you share an example that reproduces this so we can look into it?

**Comment by sacOO7** [comment]
@agiacomolli I would also suggest making sure messages are being published by checking for ack in the callback.
We recently had an issue that was closed by the customer itself due to network-related issues in such cases -> 
https://github.com/ably/ably-dotnet/issues/1252

**Comment by sacOO7** [comment]
Also, as a part of the new version, we only added support for `maxMessageSize` depending on your account limits. You can check 
https://github.com/ably/ably-go#update-maxmessagesizeread-limit-for-realtime-message-subscription for more information.

**Comment by agiacomolli** [comment]
Hi

Below is the simplest example I can get. We're using a custom endpoint for token authorization.

Logs show heartbeat messages every 15 seconds and no connection events happen (like a disconnection or failure) when messages are not received.

I'll double check the errors on web client to make sure messages are being published correctly.

```go
package main

import (
        "context"
        "log"
        "net/http"
        "os"
        "os/signal"

        "github.com/ably/ably-go/ably"
)

func main() {
        var (
                authURL     = "auth url"
                authToken   = "token"
                channelName = "channel"
        )

        authHeaders := http.Header{}
        authHeaders.Add("Authorization", authToken)

        client, err := ably.NewRealtime(
                ably.WithAuthURL(authURL),
                ably.WithAuthMethod("POST"),
                ably.WithAuthHeaders(authHeaders),
                ably.WithLogLevel(ably.LogVerbose),
        )
        if err != nil {
                log.Fatal(err)
        }

        channel := client.Channels.Get(channelName)

        ctx, cancel := signal.NotifyContext(context.Background(), os.Interrupt)
        defer cancel()

        u, err := channel.SubscribeAll(ctx, func(m *ably.Message) {
                log.Println("%v", m)
        })
        if err != nil {
                log.Fatal(err)
        }

        defer u()

        <-ctx.Done()
}
```

**Comment by sacOO7** [comment]
It will also be great if you can post js code that publishes the messages with ack. Also, check for the size of published messages. You can check https://faqs.ably.com/what-is-the-maximum-message-size for more details

**Comment by joshforbes** [comment]
Here is the publish code from the Ruby side:

```
AblyAdapter.new
      .publish(
        channel: session.agent_channel,
        name: "connect",
        data: {
          session_id: session.id,
          session_user: session.user.full_name,
          session_company: session.user.organization_name,
          sdp_offer: session.sdp_offer,
          answer_url: Routes.remote_session_answer_url(session),
          ice_servers: nts_token[:ice_servers]
        }
      )
```

And here is the AblyAdapter.publish method:

```
def publish(channel:, name:, data:)
    channel = client.channels.get(channel)
    channel.publish(name, data)
  rescue Ably::Exceptions::BaseAblyException
    false
  end
```

It's a very small message.

It isn't only this message getting lost but just one of them. If it helps, I've connected to the same channel through the Ably dev console and can see the message go through Ably while the agent completely ignores it. The message is always visible in the dev console, it's only sometimes ignored.

This is just anecdotal but it's harder to reproduce when I'm connected to the dev console - it seems less likely to happen if a second member is subscribed to the channel. When we aren't attached through the dev console, all of our channels have a single subscriber. It also seems less likely to happen when the channel is "warm", if I send a second message following the ignored one it will almost always be received.

**Comment by sacOO7** [comment]
Can you try subscribing via a different client? Because all of the information seems to be a bit vague and seems to have nothing to do with `ably-go` version right? Are you facing the same issue with old versions of `ably-go`? I would still rather ask to make sure messages are being correctly from the publisher side. For every message publishes, it should always get received from subscribers, there's no compromise there :  )

**Comment by joshforbes** [comment]
No, we didn't have this issue before we upgraded ably-go. Also, this issue only happens to a small subset of agents. Specifically, it seems most common on Linux agents - though that isn't to say that it can't happen to other platforms as well - I don't know if it can.

> I would still rather ask to make sure messages are being correctly from the publisher side.

I don't understand what you mean? The publisher side hasn't changed in years. As far as I know, that was all of the publishing code that I pasted above. It's just using Ruby to publish through HTTP. Is there something else I should be doing or something else I can show you?

>  Because all of the information seems to be a bit vague

What information should we give you to make it less vague?

**Comment by joshforbes** [comment]
Here is an example of me seeing the message in the dev console:

<img width="1254" alt="image" src="https://github.com/ably/ably-go/assets/3682518/3789cfe7-99b7-4061-981a-3fc179607425">

And here is the same timestamps on our verbose agent logs:

<img width="753" alt="image" src="https://github.com/ably/ably-go/assets/3682518/b396c787-7b6e-4c31-9b40-287e92acce54">

---

Here is what the agent logs should look like when they receive a message:

<img width="1978" alt="image" src="https://github.com/ably/ably-go/assets/3682518/097d6bc3-0095-4098-83fd-dd2be3f6c07e">

**Comment by sacOO7** [comment]
For the time being you can do 2 things
1. Either try reverting to the old version and see if that fixes the issue
2. set connection read limit explicitly to 32 kb ( if its viable ) or some other limit.

**Comment by joshforbes** [comment]
I've been digging into this all day so I haven't had a chance to review your above messages in-depth yet but I will.

I don't believe it has anything at all to do with the message size bump. I realize now that those are the versions that Anderson listed in his initial post. The version bump I was thinking of was when we went all the way from 1.13 to a modern Ably version about a month ago. I never observed the issue before we made that upgrade. However, we didn't actually notice this issue until early last week - so either it went undiagnosed for a few weeks or it's something else. I can't say at this point.

--

For now, I just wanted to post a bit more evidence. This just happened to me:

Tried to connect to a device through our UI (which sends a `connect` message). This connection failed in a manner consistent with the bug:

<img width="986" alt="image" src="https://github.com/ably/ably-go/assets/3682518/44febaea-df24-4918-af78-956b2be2e8bc">


I immediately used the Ruby API to query the channel history. The channel history included the message:

<img width="814" alt="image" src="https://github.com/ably/ably-go/assets/3682518/3360dda6-ae46-4e7f-af54-9391493aaaa9">


I had the "Iceman" agent in our verbose mode so I was able to pull the logs for the same timestamp:

<img width="1440" alt="image" src="https://github.com/ably/ably-go/assets/3682518/e6810919-488d-497c-b317-c642f55d91bf">

Unfortunately, it doesn't seem to have received the "connect" message.

**Comment by joshforbes** [comment]
A customer first reported this issue to us on July 28th. We didn't even update our staging environment to the Ably version with a message size increase until August 1st. We can rule that specific update out as a cause.

**Comment by sacOO7** [comment]
Okay, got it @joshforbes . I have created a new repo to test `ably-go` message subscription https://github.com/sacOO7/ably-go-testing. Currently, work in progress. I will update once I run the test suite against published messages.

**Comment by sacOO7** [comment]
> I've been digging into this all day so I haven't had a chance to review your above messages in-depth yet but I will.
> 
> I don't believe it has anything at all to do with the message size bump. I realize now that those are the versions that Anderson listed in his initial post. The version bump I was thinking of was when we went all the way from 1.13 to a modern Ably version about a month ago. I never observed the issue before we made that upgrade. However, we didn't actually notice this issue until early last week - so either it went undiagnosed for a few weeks or it's something else. I can't say at this point.
> 
> --
> 
> For now, I just wanted to post a bit more evidence. This just happened to me:
> 
> Tried to connect to a device through our UI (which sends a `connect` message). This connection failed in a manner consistent with the bug:
> 
> <img alt="image" width="986" src="https://user-images.githubusercontent.com/3682518/258929171-44febaea-df24-4918-af78-956b2be2e8bc.png">
> I immediately used the Ruby API to query the channel history. The channel history included the message:
> 
> <img alt="image" width="814" src="https://user-images.githubusercontent.com/3682518/258929315-3360dda6-ae46-4e7f-af54-9391493aaaa9.png">
> I had the "Iceman" agent in our verbose mode so I was able to pull the logs for the same timestamp:
> 
> <img alt="image" width="1440" src="https://user-images.githubusercontent.com/3682518/258929371-e6810919-488d-497c-b317-c642f55d91bf.png">
> Unfortunately, it doesn't seem to have received the "connect" message.

I am not sure what you mean by receiving connect message. I can clearly see from device logs that it was successfully able to connect with ably server.

**Comment by SimonWoolf** [comment]
Just to let you know: I'm a member of the Ably realtime backend team; @sacOO7 raised this issue to me and I am currently investigating to see whether this might be a serverside bug

**Comment by joshforbes** [comment]
> I am not sure what you mean by receiving connect message. I can clearly see from device logs that it was successfully able to connect with ably server.

So sorry. I see how that is confusing now! The message that we sent in this example has the name "connect". I didn't mean the actual Ably connection.

**Comment by sacOO7** [comment]
@joshforbes thanks for providing more relevant information. As @SimonWoolf mentioned, backend team is also looking into the issue, we are trying to get it resolved as fast as possible :  )

**Comment by SimonWoolf** [comment]
@joshforbes I've just rolled out a change to the backend, would appreciate if you could me whether you can still reproduce the issue now

**Comment by joshforbes** [comment]
Thanks! I'll dig in right now and let you know.

**Comment by joshforbes** [comment]
The first couple of attempts look very promising. I'm going to let the channels sit idle for a bit and then try again.

**Comment by sacOO7** [comment]
Good to hear that @joshforbes : ). Do let us know if you face any more issues.

**Comment by joshforbes** [comment]
@SimonWoolf @sacOO7 it's still working! I'll keep an eye on it but I believe you may have fixed it. Can I ask what happened?

**Comment by sacOO7** [comment]
@joshforbes I think most likely it might be a backend bug. So, nothing to do with the client as such. 
I am still testing the client from my side publishing 1K messages with 2kb size each to check if they will be received. Once the code is ready, I will post it here 👍

**Comment by sacOO7** [comment]
@joshforbes I have updated and tested the code at https://github.com/sacOO7/ably-go-testing. 
As of now, all publishers and subscribers seem to be working just fine : )

**Comment by sacOO7** [comment]
Since the issue seems to be resolved, I will close it!
You can open a new one in the future if needed :  )

**Comment by sacOO7** [comment]
@agiacomolli @joshforbes let us know if this issue is still valid https://github.com/ably/ably-go/issues/606, or got resolved as a part of #614
